rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,455 @@
+We will modify `kani-compiler` to collects stub mapping information (from the harness attributes) before code generation.
+Since stubs are specified on a per-harness basis, we need to generate multiple versions of code if all harnesses do not agree on their stub mappings; accordingly, we will update `kani-compiler` to generate multiple versions of code as appropriate. 
+To do the stubbing, we will plug in a new MIR-to-MIR transformation that replaces the bodies of specified functions with their replacements.
+This can be achieved via `rustc`'s query mechanism: if the user wants to replace `foo` with `bar`, then when the compiler requests the MIR for `foo`, we instead return the MIR for `bar`.
+`kani-compiler` will also be responsible for checking for the error conditions previously enumerated.
+
+We will also need to update the metadata that `kani-compiler` generates, so that it maps each harness to the generated code that has the right stub mapping for that harness (since there will be multiple versions of generated code).
+The metadata will also list the stubs applied in each harness.
+`kani-driver` will need to be updated to process this new type of metadata and invoke the correct generated code for each harness.
+We can also update the results report to include the stubs that were used.
+
+We anticipate that this design will evolve and be iterated upon.
+
+## Rationale and alternatives: user experience
+
+Stubbing is a *de facto* necessity for verification tools, and the lack of stubbing has a negative impact on the usability of Kani.
+
+### Benefits
+
+- Because stubs are specified by annotating the harness, the user is able to specify stubs for functions they do not have source access to (like library functions).
+This contrasts with annotating the function to be replaced (such as with function contracts).
+- The current design provides the user with flexibility, as they can specify different sets of stubs to use for different harnesses.
+This is important if users are trying to perform compositional reasoning using stubbing, since in some harnesses a function/method should be fully verified, in in other harnesses its behavior should be mocked.
+- The stub selections are located adjacent to the harness, which makes it easy to understand which replacements are going to happen for each harness.
+
+### Risks
+
+- Users can always write stubs that do not correctly correspond to program behavior, and so a successful verification does not actually mean the program is bug-free.
+This is similar to other specification bugs.
+All the stubbing code will be available, so it is possible to inspect the assumptions it makes.
+
+### Comparison to function contracts
+
+- The [currently proposed function contract mechanism](https://github.com/model-checking/kani/tree/features/function-contracts) does not provide a way to specify contracts on external functions.
+This is one of the key motivations for stubbing.
```

Comment:
I wouldn't say this is the key motivation because conceptually, it should be possible to allow contracts to be specified on external functions (which AFAIK, CBMC allows), but it hasn't been prioritized in the current implementation. I would say the key benefit of stubbing is the simplicity of replacing a function's implementation with another that is written in Rust. The replacement can be really be anything: an over-approximation, an under-approximation, or neither. On the other hand, the disadvantage of stubbing is that there's no easy mechanism to verify that a proof is sound with the specified stubs.